rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,710 @@
+behavior in the generated documentation, but in other tools as well, such as automatically
+transforming one version of a configuration to a newer version by analyzing the schemas, and being
+able to reason that if field X used to be able to be referenced via A or B, and now only B is
+allowed, we can just rename A to B. We could also check that schemas don't remove fields unless
+those fields were already marked as deprecated, being able to enforce our unofficial guideline of
+not removing fields unless they've been marked as deprecated for at least one release.
+
+#### 4. Utilize `schemars` to generate the actual JSONSchema
+
+With the necessary information being derived from our configuration types directly, we still need a
+way to take that and emit an actual JSONSchema document for Vector's configuration. We would utilize
+[`schemars`][schemars], a crate for generating JSONSchema documents, for this purpose.
+
+The `schemars` crate, among other things, provides a small set of traits and types related
+specifically to programmatically generating a JSONSchema document. Types wish a user wishes to
+document must implement `JSONSchema`, which interacts with a `SchemaGenerator` object that actually
+holds the in-progress schema as a type is being walked.
+
+Our `Configurable` derive macro would implement this trait automatically for the given type, using
+the information provided by `Configurable` to ultimately drive the schema generation.
+
+## Rationale
+
+In general, the lack of a configuration schema, and being able to treat our Rust code as the source
+of truth, hurts both users and developers. As such, doing this work represents a huge opportunity to
+reduce developer toil when it comes to generating and synchronizing our generated documentation.  It
+additionally represents a large quality of life improvement for our users, who look to our
+documentation to be up-to-date and semantically meaningful.
+
+If we didn't do this work, we would be stuck with the current manual toil, which is not only extra
+work on the part of developers -- learning Cue, remembering to update it, etc -- but also reviewers,
+who now need to catch when these things are missed.
+
+Utilizing an extensible system for expressing the schema of Vector's configuration gives us the
+runway to solve our current problems, but also the ability to handle future problems and
+requirements as they come up.
+
+## Drawbacks
+
+The primary drawback of this approach is that it limits us to things we can reasonably express
+within the limitations of the Rust language itself, and what we're comfortable with representing via
+attributes. For example, single-line and multi-line doc comments are trivially supported, but if we
+wanted to start pushing more semantically-relevant information, such as configuration snippets, etc,
+it could be technically posssible but appear as a very ugly attribute usage, which could make the
+source muddy and unclear.
```

Comment:
For context, we are planning to add more examples to the `config/examples` folder that'll align with use cases and guides. I plan to reference related use cases in the component's documentation, so I don't think real-world examples need to be embedded into the code.

Reply:
Fair enough.

My personal inclination is that we have many sources where they use something like the "string or object" pattern, and it's not on a common/shared configuration block (like buffer or batch config) and so it might be hard/unwieldy to generate full examples for each of those.. which is where inline comments that could hold both description and example(s) could become relevant and useful.

Reply:
Got it. Yeah, we'll need to figure that out in the interim but, ideally, we'd eliminate polymorphism entirely in our schema. That's what we were [going for in the configuration spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/configuration.md#polymorphism). I expect this to be a follow up project once we figure out how to more easily support backward compatibility between versions.